src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c

```diff
@@ -0,0 +1,986 @@
+			ert_user->submit_queue[i] = NULL;
+			ert_user_free_cmd(ecmd);
+			--ert_user->num_sq;
+		}
+	}
+	spin_unlock_irqrestore(&ert_user->sq_lock, flags);
+}
+/*
+ * release_slot_idx() - Release specified slot idx
+ */
+static void
+ert_release_slot_idx(struct xocl_ert_user *ert_user, unsigned int slot_idx)
+{
+	clear_bit(slot_idx, ert_user->slot_status);
+}
+
+/**
+ * release_slot() - Release a slot index for a command
+ *
+ * Special case for control commands that execute in slot 0.  This
+ * slot cannot be marked free ever.
+ */
+static void
+ert_release_slot(struct xocl_ert_user *ert_user, struct ert_user_command *ecmd)
+{
+	if (ecmd->slot_idx == no_index)
+		return;
+
+	if (cmd_opcode(ecmd) == OP_CONFIG) {
+		ERTUSER_DBG(ert_user, "do nothing %s\n", __func__);
+		ert_user->ctrl_busy = false;
+		ert_user->config = true;	
+	} else {
+		ert_release_slot_idx(ert_user, ecmd->slot_idx);
+	}
+	ecmd->slot_idx = no_index;
+}
+
+/**
+ * process_ert_cq() - Process completed queue
+ * @ert_user: Target XRT CU
+ */
+static inline void process_ert_cq(struct xocl_ert_user *ert_user)
+{
+	struct kds_command *xcmd;
+	struct ert_user_command *ecmd;
+
+	if (!ert_user->num_cq)
+		return;
+
+	ERTUSER_DBG(ert_user, "-> %s\n", __func__);
+	/* Notify host and free command */
+	ecmd = list_first_entry(&ert_user->cq, struct ert_user_command, list);
+	xcmd = ecmd->xcmd;
+	ERTUSER_DBG(ert_user, "%s -> ecmd->slot_idx %d xcmd%p\n", __func__, ecmd->slot_idx, xcmd);
+	ert_release_slot(ert_user, ecmd);
+	list_del(&ecmd->list);
+	xcmd = ecmd->xcmd;
+	xcmd->cb.notify_host(xcmd, KDS_COMPLETED);
+	xcmd->cb.free(xcmd);
+	ert_user_free_cmd(ecmd);
+	--ert_user->num_cq;
+	ERTUSER_DBG(ert_user, "<- %s\n", __func__);
+}
+
+/**
+ * mask_idx32() - Slot mask idx index for a given slot_idx
+ *
+ * @slot_idx: Global [0..127] index of a CQ slot
+ * Return: Index of the slot mask containing the slot_idx
+ */
+static inline unsigned int
+mask_idx32(unsigned int idx)
+{
+	return idx >> 5;
+}
+
+static irqreturn_t
+ert_user_versal_isr(int irq, void *arg)
+{
+	struct xocl_ert_user *ert_user = (struct xocl_ert_user *)arg;
+	xdev_handle_t xdev = xocl_get_xdev(ert_user->pdev);
+
+	ERTUSER_DBG(ert_user, "-> %s %d\n", __func__, irq);
+
+	if (ert_user) {
+		xocl_mailbox_versal_handle_intr(xdev);
+#if 0
+		if (!ert_user->polling_mode)
+			scheduler_intr(exec->scheduler);
+		else
+			ERTUSER_ERR(ert_user, "unhandled isr irq %d", irq);
+#endif
+	}
+
+	ERTUSER_DBG(ert_user, "<- %s\n", __func__);
+	return IRQ_HANDLED;
+}
+
+static irqreturn_t
+ert_user_isr(int irq, void *arg)
+{
+	struct xocl_ert_user *ert_user = (struct xocl_ert_user *)arg;
+	xdev_handle_t xdev = xocl_get_xdev(ert_user->pdev);
+	struct ert_user_command *ecmd;
+
+	ERTUSER_DBG(ert_user, "-> xocl_user_event %d\n", irq);
+	/*
+	 * versal_isr is registered here,
+	 * but versal interrupt is enabled by mailbox_versal subdev.
+	 * Note: should be separated from ert_user_isr in new scheduler.
+	 */
+
+	if (ert_user && XOCL_DSA_IS_VERSAL(xdev))
```

Comment:
Let's talk with @larry9523 to figure out how to avoid use XOCL_DSA_IS_VERSAL().
But for this pull request, I think this is okay.

Reply:
After discussed with @chienwei-lan and @larry9523, INTC sub-device should deal with the interrupt implementation difference. This ert_user sub-device should always see 128 logical interrupts. 
I'd enhance INTC sub-device once this PR is merged.